### PR TITLE
Signins and signouts work.

### PR DESCRIPTION
**Pull request by betweenparentheses**

### CODE REVIEW

File: app/controllers/admin/addresses_controller.rb [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+  def check_city
+    city = params[:address][:city]
+
+    if City.find_by(name: city)
+      params[:address][:city_id] = City.find_by(name: city).id
+    else
+      c = City.create(name: city)
+      params[:address][:city_id] = c.id
+    end
```
Comment (behaviary) [inline app/controllers/admin/addresses_controller.rb:81]: I like check_city. This is a great solution for simplifying things in the update and create methods.

File: app/controllers/admin/categories_controller.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      flash[:success] = "Updated!"
+      redirect_to action: :index
+    else
+      flash.now[:error] = "Something went wrong."
+      render :edit
+    end
+  end
+
+  def destroy
+    session[:return_to] ||= request.referer
```
Comment (behaviary) [inline app/controllers/admin/categories_controller.rb:45]: interesting choice for redirecting. This takes them back to whatever page they came from, right?

File: app/controllers/admin/dashboard_controller.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
+class Admin::DashboardController < AdminController
+  def index
+    @all_time = nil
+
+    #add extra days and it auto-generates new tables via partial
+    @time_frames = [7, 30, @all_time]
```
Comment (behaviary) [inline app/controllers/admin/dashboard_controller.rb:6]: This is cool. I remember when y'all demoed it. Still cool.

File: app/controllers/admin/orders_controller.rb [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+    #special hidden params to ID which edit form was used
+    update_purchases if params[:order][:update]
+    add_contents if params[:order][:add]
+
+    if @order.update(whitelisted_params) || @order.purchases.each {|p| p.save}
+      flash[:success] = "Updated!"
+      redirect_to [:admin, @order]
+    else
+      flash.now[:error] = "Failed to save due to #{@order.errors.full_messages.inspect}\n #{@order.purchases.last.errors.full_messages.inspect}"
```
Comment (behaviary) [inline app/controllers/admin/orders_controller.rb:48]: nice nested errors

File: app/controllers/admin/purchases_controller.rb [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        render :new
+      end
+    end
+
+    def show
+      @purchase = Purchase.find(params[:id])
+
+      @purchase_numorders = @purchase.orders.count
+      @purchase_numcarts = @purchase.orders.where(:checked_out => false).count
+    end
```
Comment (behaviary) [inline app/controllers/admin/purchases_controller.rb:28]: the Purchase show method looks familiar... Helper for counting orders?

File: app/controllers/admin/orders_controller.rb [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+      redirect_to action: :index
+    else
+      flash[:error] = "Something went wrong in that deletion."
+      redirect_to session.delete(:return_to)
+    end
+  end
+
+  private
+
+  def add_contents
```
Comment (behaviary) [inline app/controllers/admin/orders_controller.rb:66]: This is a beast! I bet you'd write this very differently after this week :)

Comment (betweenparentheses) [inline app/controllers/admin/orders_controller.rb:66]: MONDAY MADE ME SO ANGRY HAHAHAHAHAHAHA

On Thu, Oct 16, 2014 at 12:01 PM, Peter notifications@github.com wrote:

> In app/controllers/admin/orders_controller.rb:
> 
> > -  end
> >   +
> > -  def destroy
> > -    session[:return_to] ||= request.referer
> > -    if Order.find(params[:id]).destroy
> > -      flash[:success] = "Deleted."
> > -      redirect_to action: :index
> > -    else
> > -      flash[:error] = "Something went wrong in that deletion."
> > -      redirect_to session.delete(:return_to)
> > -    end
> > -  end
> >   +
> > -  private
> >   +
> > -  def add_contents
> 
> This is a beast! I bet you'd write this very differently after this week :)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/vikingeducation/viking_store_admin/pull/1/files#r18978172
> .

## 

_Michael Alexander_

(310) 429-8822

alexandermw@gmail.com

File: app/controllers/admin/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    def whitelisted_params
+      params.require(:user).permit(:first_name, :last_name,
+                                   :email, :billing_id, :shipping_id)
+    end
+
+    #takes a string parameter
+    def clean_price(input)
+      input[0] == "$" ? input[1..-1].to_f : input.to_f
+    end
```
Comment (behaviary) [inline app/controllers/admin/users_controller.rb:68]: Love that! never trust your users.

**Comment by behaviary** [comment]
Overall: very well done!  Honestly, I feel a bit under-qualified to review your code. :smile: 
Things I didn't add line comments for: Your views are very clean! I think you guys did a really good job at keeping logic out of there.  The controllers looked good except for some minor things that I know you already know how to fix.  I think that there were a couple places where you could have used helpers to make your life easier, but again, they were nominal. There are two things I was actually unsure of: 
1. this specific line: 
   **app/controllers/admin/orders_controller.rb
   44 +    if @order.update(whitelisted_params) || @order.purchases.each {|p| p.save}
2. Your approach to redirecting is definitely different than mine

I want to ask you about both of dem thangs when we are working together.

In conclusion, you're a rockstar. Keep up the amazing work!

**Comment by betweenparentheses** [comment]
Thanks!

#1) Actually completely sucks and is in no way good code. You are right to
go "WTF"??  Basically, it was me throwing up my hands and saying "fuck it"
because i didn't have the energy to fake nested parameters once I found out
how, and so I couldn't just save purchases using .build and saving the
parent (because Erik is a forms and associations sadist laughing at our
tears).  Basically, this is always true and makes weird errors very
possible, but it does manage to save all the savable purchases as a side
effect of updating the order. CODING SHAME.

2) There are so many different ways to redirect, I have no idea what's
normal yet (using url_for? action? path helpers? resources? they all seem
to work, and I basically seem to be cycling through them just to know what
they all are.). Let's definitely talk it out and figure out an approach
that works.

This week is extra insane, right? Even more than normal? But I definitely
believe that we can knock this one cold once we get into it this weekend.

On Thu, Oct 16, 2014 at 12:14 PM, Peter notifications@github.com wrote:

> Overall: very well done! Honestly, I feel a bit under-qualified to review
> your code. [image: :smile:]
> Things I didn't add line comments for: Your views are very clean! I think
> you guys did a really good job at keeping logic out of there. The
> controllers looked good except for some minor things that I know you
> already know how to fix. I think that there were a couple places where you
> could have used helpers to make your life easier, but again, they were
> nominal. There are two things I was actually unsure of:
> 
>    1.
> 
>    this specific line:
>    app/controllers/admin/orders_controller.rb
> 
>    44 + if @order.update(whitelisted_params) || @order.purchases.each
>    {|p| p.save}
>     2.
> 
>    Your approach to redirecting is definitely different than mine
> 
> I want to ask you about both of dem thangs when we are working together.
> 
> In conclusion, you're a rockstar. Keep up the amazing work!
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/vikingeducation/viking_store_admin/pull/1#issuecomment-59414496
> .

## 

_Michael Alexander_

(310) 429-8822

alexandermw@gmail.com
